oauth2/fosite_store_sql.go

```diff
@@ -447,7 +447,18 @@ func (s *FositeSQLStore) FlushInactiveAccessTokens(ctx context.Context, notAfter
 }
 
 func (s *FositeSQLStore) BeginTX(ctx context.Context) (context.Context, error) {
-	if tx, err := s.DB.BeginTxx(ctx, nil); err != nil {
+	// TL;DR make it a no-op if the transaction has already been initiated:
+	// some 'fosite' APIs may attempt to initiate a transaction and they may not be aware that
+	// one has already been initiated by the caller so we make this a no-op in those cases where
+	// the caller (hydra) has more knowledge of the transaction boundary.
+	if _, ok := ctx.Value(txKey).(*sqlx.Tx); ok {
+		return ctx, nil
+	}
+
+	if tx, err := s.DB.BeginTxx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
```

Comment:
If you change this to `sql.LevelReadCommitted` and run the test scenario added, you will be able to successfully use the same refresh token [more than once](https://github.com/ory/hydra/pull/1766/files#diff-8a4c55f53fc1d3a445cf9930b67cfafaR149) 😨

Reply:
Nice find man!